fastplotlib/legends/legend.py

```diff
@@ -0,0 +1,344 @@
+        return tuple(self._graphics)
+
+    def _check_label_unique(self, label):
+        for legend_item in self._items.values():
+            if legend_item.label == label:
+                raise ValueError(
+                    f"You have passed the label '{label}' which is already used for another legend item. "
+                    f"All labels within a legend must be unique."
+                )
+
+    def add_graphic(self, graphic: Graphic, label: str = None):
+        if graphic in self._graphics:
+            raise KeyError(
+                f"Graphic already exists in legend with label: '{self._items[graphic.loc].label}'"
+            )
+
+        self._check_label_unique(label)
+
+        new_col_ix = self._col_counter
+        new_row_ix = self._row_counter
+
+        x_pos = 0
+        y_pos = 0
+
+        if self._row_counter == self._max_rows:
+            # set counters
+            new_col_ix = self._col_counter + 1
+
+            # get x position offset
+            # get largest x_val from bbox of previous column bboxes
+            prev_column_items: List[LegendItem] = list(self._items.values())[-self._max_rows:]
+            x_pos = prev_column_items[-1].world_object.world.x
+            max_width = 0
+            for item in prev_column_items:
+                bbox = item.world_object.get_world_bounding_box()
+                width, height, depth = bbox.ptp(axis=0)
+                max_width = max(max_width, width)
+            x_pos = x_pos + max_width + 15  # add 15 for spacing
+
+            # rest row index for next iteration
+            new_row_ix = 1
+        else:
+            if len(self._items) > 0:
+                x_pos = list(self._items.values())[-1].world_object.world.x
+
+            y_pos = new_row_ix * -10
+            new_row_ix = self._row_counter + 1
+
+        if isinstance(graphic, LineGraphic):
+            legend_item = LineLegendItem(self, graphic, label, position=(x_pos, y_pos))
+        else:
+            raise ValueError("Legend only supported for LineGraphic for now.")
+
+        self._legend_items_group.add(legend_item.world_object)
+        self._reset_mesh_dims()
+
+        self._graphics.append(graphic)
+        self._items[graphic.loc] = legend_item
+        graphic.deleted.add_event_handler(partial(self.remove_graphic, graphic))
+
+        self._col_counter = new_col_ix
+        self._row_counter = new_row_ix
+
+    def _reset_mesh_dims(self):
+        bbox = self._legend_items_group.get_world_bounding_box()
+
+        width, height, _ = bbox.ptp(axis=0)
+
+        self._mesh.geometry.positions.data[mesh_masks.x_right] = width + 7
+        self._mesh.geometry.positions.data[mesh_masks.x_left] = -5
+        self._mesh.geometry.positions.data[mesh_masks.y_bottom] = 0
+        self._mesh.geometry.positions.data[mesh_masks.y_bottom] = -height - 3
+        self._mesh.geometry.positions.update_range()
+
+    def remove_graphic(self, graphic: Graphic):
```

Comment:
Does deleting a graphic from the plot automatically remove it from the legend if it is in the legend?

Reply:
yes, see: https://github.com/fastplotlib/fastplotlib/pull/406/files#diff-042597b7ed37138aa42fa3fdbdf41423475a26bbb17c0d585b8449a0c0419ad9R271

Reply:
ahh, always the partials